### DISCUSSION TITLE
How can I get the calling name of the method which called .SaveChangesAsync();

### DISCUSSION THREAD

**Discussion by jackdh** [thepirat000/Audit.NET#567] [Q&A]
I have the following Setup

```C#
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .AuditTypeExplicitMapper(m => m
            .Map<InputRow, AuditLog>()
            .Map<GreenLightSystem, AuditLog>()
            .AuditEntityAction<AuditLog>((evt, entry, auditEntity) =>
            {
                auditEntity.AuditData = entry.ToJson();
                auditEntity.AuditDate = DateTime.UtcNow;
                auditEntity.CallingMethod = "Calling Method"; <-- Where I need the calling method.
            })
        )
    );
```

How can I get access to the method which called .saveChangesAsync so I know where it was triggered from? I've checked using StackFrame however that is flooded with library level frames which are not very helpful.

I see in evt that it provides a calling parent name, however that is just one level up so is just the "SaveChangesAsync" method.

Thanks

**Comment by thepirat000** [comment]
yeah, the problem is that every combination of  EntityFramework version, .NET Core version, build configuration (debug / release) and compiler optimizations will have a different stack traces when the Audit Scope is created, so it's almost impossible for the library to tell how many frames to go up with certainly. 

Here is my best attempt to guess the frames that are needed to skip to get to the method that called SaveChanges, but as you see it doesn't always work:

https://github.com/thepirat000/Audit.NET/blob/fd085ce75f009fc2a90752dea9d31bed7af9d097/src/Audit.EntityFramework/DbContextHelper.cs#L319

So I was thinking that maybe the "SkipExtraFrames" value should be globally configurable (it's currently configurable when you create the scope, but the scope is created behind the scenes by the EF extension), so you could define the value that works specifically for your use case. But again it will not be 100% reliable, since any change on the versions will mostly change the frame position in the stack trace.

**Reply by jackdh** [reply]
Thanks mate I'll check that out.


I was thinking I could perhaps make an extension on the .saveChangesAsync() which would require a param of the calling method, then is there a way to set that as a custom field? I'm sorry brand new to your great project!

**Reply by thepirat000** [reply]
That's a great idea. 

Yes, you can add custom fields within your DbContext with `AddAuditCustomField()` method.

For example:

```c#
public class YourContext : AuditDbContext
{
    // ...

    public override int SaveChanges()
    {
        throw new NotImplementedException();
    }

    public override Task<int> SaveChangesAsync(CancellationToken cancellationToken = new CancellationToken())
    {
        throw new NotImplementedException();
    }

    public int SaveChangesAudit([CallerMemberName] string memberName = "", [CallerFilePath] string filePath = "", [CallerLineNumber] int line = 0)
    {
        this.AddAuditCustomField("Caller", new { memberName, filePath, line });
        return base.SaveChanges();
    }
    public Task<int> SaveChangesAuditAsync([CallerMemberName] string memberName = "", [CallerFilePath] string filePath = "", [CallerLineNumber] int line = 0)
    {
        this.AddAuditCustomField("Caller", new { memberName, filePath, line });
        return base.SaveChangesAsync();
    }
}
```

So your code should now call the `SaveChangesAudit()` / `SaveChangesAuditAsync()` methods.

In your audit event you will see something like this:

```javascript
{
  "EntityFrameworkEvent": {
     ...
  },
  "Caller": {
    "memberName": "InsertAsync",
    "filePath": "C:\\Temp\\eftest\\Services\\ValuesService.cs",
    "line": 35
  }
}
```

**Reply by jackdh** [reply]
Nice that's working well! Thank you.

I do have a concern however which might be invalid that data here could be corrupted incase of a race condition with multiple SaveChangesAsync going in? In my current case that's not too much of an issue but if I was to scale this up it might be?

**Comment by thepirat000** [comment]
The EF DbContext is not threadsafe, so you shouldn't have multiple threads calling SaveChanges on the same instance
